File: C _ C++/C++/DSA/Search In Array/BinarySearch.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int n, int target)
{
    int s = 0;
    int e = n;
    while (s <= e)
    {
        int mid = (s + e) / 2;
        if (arr[mid] == target)
        {
            return mid;
        }
        else if (arr[mid] > target)
        {
            e = mid - 1;
        }
        else
        {
            s = mid + 1;
        }
    }
    return -1;
}

int main()
{
    int n;
    cin >> n;
    int arr[n], target;
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    cin >> target;
    cout << binarySearch(arr, n, target) << endl;
    return 0;
}

File: Algorithms/level2/Search/binarySearch.py
# Iterative binary search

def binary_search(arr, start, end, target):
    while start <= end:
        mid = (start+end)//2
        if arr[mid] < target:
            start = mid + 1
        elif arr[mid] > target:
            end = mid - 1
        else :
            return mid
    return start

arr = [2, 5, 8, 10, 11, 22, 25]
target = 8

result = binary_search(arr, 0, len(arr)-1, target)
# Recursive method

def BinarySearch(arr, start, end, target):
    if end >= start:
        mid = (start+end)//2
        if arr[mid] < target:
            x = BinarySearch(arr, mid+1, end, target)
        elif arr[mid] > target:
            x = BinarySearch(arr, start, mid-1, target)
        else :
            return mid
    else :
        return -1
    return x

arr = [2, 5, 8, 10, 11, 22, 25]
target = 8

result = BinarySearch(arr, 0, len(arr)-1, target)
print(result)